libfulltext/arxiv/__init__.py

```diff
@@ -0,0 +1,73 @@
+# copyright © 2018 the libfulltext authors (see AUTHORS.md and LICENSE)
+"""arXiv handler module"""
+
+import requests
+from lxml import etree
+
+from ..response import verify
+from ..doi import get_doi_fulltext
+
+
+def get_arxiv_fulltext(arxiv_id, save_stream, config):
+    """Get the fulltext for an arXiv ID
+
+    At the moment meant to handle one arXiv ID at a time, though the arxiv
+    API can handle multiple IDs (comma separated).
+
+    Args:
+        arxiv_id:    arxiv ID string (with prefix)
+        save_stream: function that saves a stream (arguments: stream, path)
+        config:      configuration dictionary (see config.py)
+
+    Returns:
+        What the actual getter returns (usually None)
+
+    Raises:
+        ValueError:           did not obtain exactly one query result or
+                              entry did not contain pdf link
+        NotImplementedError:  multiple arXiv IDs provided
+    """
+
+    if "," in arxiv_id:
+        raise NotImplementedError("multiple arXiv IDs not implemented")
+
+    response = requests.get("http://export.arxiv.org/api/query",
+                            params={"id_list": arxiv_id})
+    tree = etree.fromstring(response.content)
+
+    # './' = restrict search to depth 1 in the tree
+    # '{*}' = ignore the namespace (usually "http://www.w3.org/2005/Atom")
+    entries = tree.findall("./{*}entry")
+
+    if len(entries) > 1:
+        raise ValueError("Obtained more than one arXiv article")
+    elif len(entries) == 0:
+        raise ValueError("Did not obtain any arXiv article")
+
+    entry = entries[0]
+
+    pdf_links = []
+    for element in entry.findall("./{*}link"):
+        if 'title' in element.keys() and element.get("title") == "pdf":
+            pdf_links.append(element.get("href"))
+
+    if len(pdf_links) == 0:
+        raise ValueError("Didn't contain a pdf link")
+    elif len(pdf_links) > 1:
+        raise ValueError("Ambiguous pdf links")
```

Comment:
In the elsevier fetcher we use a `HTTPError` from `requests` for these kind of issues. I am not sure whether `HTTPError` is the right thing to do, but a `ValueError` certainly feels wrong, since is not an issue of the input side, right?
 
As a side note: We maybe should think about custom excepions for this. In `python` a custom exception can be as simple as declaring the name and a docstring (see the `ConfigurationError` exception in the `feature/config-env-vars` branch).

Reply:
Agree. It seems to me a few custom exceptions (for libfulltext, not just arxiv) make sense.